### ISSUE TITLE
Bias Variance Tradeoff

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8834]
Bias Variance Tradeoff<br>
<br>
https://ift.tt/wIzLQGp<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<article>
<div>
<p>Using our wildest imagination, we can picture a dataset consisting of features X and labels Y, as on the left. Also imagine that we&rsquo;d like to generalize this relationship to additional values of X - that we&rsquo;d like to predict future values based on what we&rsquo;ve already seen before.<br>
<br>
With our imagination now undoubtedly spent, we can take a very simple approach to modeling the relationship between X and Y by just drawing a line to the general trend of the data.</p>
</div>
<div>
<p>A Simple Model</p>
<p>Our simple model isn&rsquo;t the best at modeling the relationship - clearly there's information in the data that it's failing to capture.<br>
<br>
We'll measure the performance of our model by looking at the <a href="https://en.wikipedia.org/wiki/Mean_squared_error">mean-squared error</a> of its output and the true values (displayed in the bottom barchart). Our model is close to some of the training points, but overall there's definitely room for improvement.<br>
<br>
The error on the training data is important for model tuning, but what we really care about is how it performs on data we haven't seen before, called test data. So let's check that out as well.</p>
</div>
<div>
<p>Low Complexity &amp; Underfitting</p>
<p>Uh-oh, it looks like our earlier suspicions were correct - our model is garbage. The test error is even higher than the train error!<br>
<br>
In this case, we say that our model is <span>underfitting</span> the data: our model is so simple that it fails to adequately capture the relationships in the data. The high test error is a direct result of the lack of complexity of our model.<br>
<br>
<span>An underfit model is one that is too simple to accurately capture the relationships between its features X and label Y.</span></p>
</div>
<div>
<p>A Complex Model</p>
<p>Our previous model performed poorly because it was too simple. Let's try our luck with something more complex. In fact, let's get as complex as we can - let's train a model that predicts every point in our training data perfectly.<br>
<br>
Great! Now our training error is zero. As the old saying goes in Tennessee: Fool me once - shame on you. Fool me twice - er... you can't get fooled again ;).</p>
</div>
<div>
<p>High Complexity &amp; Overfitting</p>
<p>Wait a second... Even though our training error from our model was effectively zero, the error on our test data is high. What gives?<br>
<br>
Unsurprisingly, our model is too complicated. We say that it <span>overfits</span> the data. Instead of learning the true trends underlying our dataset, it memorized noise and, as a result, the model is not generalizable to datasets beyond its training data.<br>
<br>
<span>Overfitting refers to the case when a model is so specific to the data on which it was trained that it is no longer applicable to different datasets.</span><br>
<br>
In situations where your training error is low but your test error is high, you've likely overfit your model.</p>
</div>
<div>
<p>Test Error Decomposition</p>
<p>Our test error can come as a result of both under- and over- fitting our data, but how do the two relate to each other?<br>
<br>
In the general case, <span>mean-squared error can be decomposed into three components: error due to bias, error to to variance, and error due to noise.</span><br>
<br>
<br>
<br>
Or, mathematically:<br>
<br>
<br>
<br>
We can&rsquo;t do much about the irreducible term, but we can make use of the relationship between both bias and variance to obtain better predictions.</p>
</div>
<div>
<p>Bias</p>
<p><span>Bias represents the difference between the average prediction and the true value</span>:<br>
<br>
<br>
<br>
The term is a tricky one. It refers to the average prediction after the model has been trained over several independent datasets. We can think of the bias as measuring a <i>systematic</i> error in prediction.<br>
<br>
These different model realizations are shown in the top chart, while the error decomposition (for each point of data) is shown in the bottom chart.<br>
<br>
For underfit (low-complexity) models, the majority of our error comes from bias.</p>
</div>
<div>
<p>Variance</p>
<p>As with bias, the notion of variance also relates to different realizations of our model. Specifically, <span>variance measures how much, on average, predictions vary for a given data point</span>:<br>
<br>
<br>
<br>
As you can see in the bottom plot, predictions from overfit (high-complexity) models show a lot more error from variance than from bias. It&rsquo;s easy to imagine that any unseen data points will be predicted with high error.</p>
</div>
<div>
<p>Finding A Balance</p>
<p>To obtain our best results, we should work to find a happy medium between a model that is so basic it fails to learn meaningful patterns in our data, and one that is so complex it fails to generalize to unseen data .<br>
<br>
In other words, we don&rsquo;t want an underfit model, but we don&rsquo;t want an overfit model either. We want something in between - something with enough complexity to learn learn the generalizable patterns in our data.<br>
<br>
<span>By trading some bias for variance (i.e. increasing the complexity of our model), and without going overboard, we can find a balanced model for our dataset</span>.</p>
</div>
<div>
<p>Across Complexities</p>
<p>We just showed, at different levels of complexity, a sample of model realizations alongside their corresponding prediction error decompositions.<br>
<br>
Let&rsquo;s direct our focus to the error decompositions across model complexities.<br>
<br>
For each level of complexity, we&rsquo;ll aggregate the error decomposition across all data-points, and plot the aggregate errors at their level of complexity.<br>
<br>
This aggregation applied to our balanced model (i.e. the middle level of complexity) is shown to the left.</p>
</div>
<div>
<p>The Bias Variance Trade-off</p>
<p>Repeating this aggregation across our range of model complexities, we can see the relationship between bias and variance in prediction errors manifests itself as a U-shaped curve detailing the trade off between bias and variance.<br>
<br>
When a model is too simple (i.e. small values along the x-axis), it ignores useful information, and the error is composed mostly of that from bias.<br>
<br>
When a model is too complex (i.e. large values along the x-axis), it memorizes non-general patterns, and the error is composed mostly of that from variance.<br>
<br>
<span>The ideal model aims to minimize both bias and variance. It lays in the sweet spot - not too simple, nor too complex. Achieving such a balance will yield the minimum error.</span></p>
</div>
</article>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live.<br>
<br>
September 21, 2022 at 12:09AM
